File: MacTemplet/Cagtegory/NSTextView+Helper.m
//
//  NSTextView+Helper.m
//  MacTemplet
//
//  Created by Bin Shang on 2019/6/10.
//  Copyright © 2019 Bin Shang. All rights reserved.
//

#import "NSTextView+Helper.h"

@implementation NSTextView (Helper)

static NSDictionary *_dic;

+ (NSDictionary *)dic{
    if (!_dic) {
        _dic = @{
                 @"x": @"cut:",
                 @"c": @"copy:",
                 @"v": @"paste:",
                 @"a": @"selectAll:",
                 @"z": @"keyDown:",
                 };
    }
    return _dic;
}

- (BOOL)performKeyEquivalent:(NSEvent *)event{
    if ((event.modifierFlags & NSEventModifierFlagDeviceIndependentFlagsMask) == NSEventModifierFlagCommand) {
        // The command key is the ONLY modifier key being pressed.
//        SEL selector = NSSelectorFromString(NSTextView.dic[event.charactersIgnoringModifiers]);
//        return [NSApp sendAction:selector to: self.window.firstResponder from:self];
        if ([event.charactersIgnoringModifiers isEqualToString:@"x"]) {
            return [NSApp sendAction:@selector(cut:) to: self.window.firstResponder from:self];
        } else if ([event.charactersIgnoringModifiers isEqualToString:@"c"]) {
            return [NSApp sendAction:@selector(copy:) to:self.window.firstResponder from:self];
        } else if ([event.charactersIgnoringModifiers isEqualToString:@"v"]) {
            return [NSApp sendAction:@selector(paste:) to:self.window.firstResponder from:self];
        } else if ([event.charactersIgnoringModifiers isEqualToString:@"a"]) {
            return [NSApp sendAction:@selector(selectAll:) to:self.window.firstResponder from:self];
        } else if ([event.charactersIgnoringModifiers isEqualToString:@"z"]) {
//            return [NSApp sendAction:@selector(keyDown:) to:self.window.firstResponder from:self];
            
        }
    }
    return [super performKeyEquivalent:event];
}

-(void)setHyperlinkDic:(NSDictionary *)dic{
    // both are needed, otherwise hyperlink won't accept mousedown
    NSTextView *textView = self;
    
    NSDictionary * attributes = @{
                                  NSFontAttributeName: textView.font,

                                  };
    
    NSAttributedString * attStr = [[NSAttributedString alloc]initWithString:textView.string attributes:attributes];
    
    __block NSMutableAttributedString * mattStr = [[NSMutableAttributedString alloc]init];
    [mattStr replaceCharactersInRange:NSMakeRange(0, 0) withAttributedString:attStr];
    [dic enumerateKeysAndObjectsUsingBlock:^(NSString * key, id  _Nonnull obj, BOOL * _Nonnull stop) {
        NSURL *url = [NSURL URLWithString:obj];
        NSAttributedString * attStr = [NSAttributedString hyperlinkFromString:key withURL:url font:textView.font];
        NSRange range = [mattStr.string rangeOfString:key];
        [mattStr replaceCharactersInRange:range withAttributedString:attStr];
        
    }];
    [textView.textStorage setAttributedString:mattStr];

    textView.editable = false;
    textView.selectable = true;
    
}
@end

File: MacTemplet/CocoaExpand/NSTextView+Helper.swift
//
//  NSTextView+Helper.swift
//  MacTemplet
//
//  Created by Bin Shang on 2019/11/23.
//  Copyright © 2019 Bin Shang. All rights reserved.
//

import Cocoa

@objc public extension NSTextView {
    
    override func performKeyEquivalent(with event: NSEvent) -> Bool {
        if event.modifierFlags.isDisjoint(with: .command) {
            return super.performKeyEquivalent(with: event)
        }
        
        switch event.charactersIgnoringModifiers {
        case "a":
            return NSApp.sendAction(#selector(NSText.selectAll(_:)), to: self.window?.firstResponder, from: self)
        case "c":
            return NSApp.sendAction(#selector(NSText.copy(_:)), to: self.window?.firstResponder, from: self)
        case "v":
            return NSApp.sendAction(#selector(NSText.paste(_:)), to: self.window?.firstResponder, from: self)
        case "x":
            return NSApp.sendAction(#selector(NSText.cut(_:)), to: self.window?.firstResponder, from: self)
        case "z":
            self.window?.firstResponder?.undoManager?.undo()
            return true
        case "Z":
            self.window?.firstResponder?.undoManager?.redo()
            return true
        default:
            return super.performKeyEquivalent(with: event)
        }
    }
    
    static func create(_ rect: CGRect) -> Self {
        let view: NSTextView = self.init(frame: rect);
        view.autoresizingMask = [.width, .height];
           
        view.font = NSFont.systemFont(ofSize: 14)
        view.textColor = NSColor.black;

        view.isHorizontallyResizable = false;
        view.isVerticallyResizable = true;

        view.maxSize = CGSize(width: CGFloat.greatestFiniteMagnitude, height: CGFloat.greatestFiniteMagnitude)
        view.isSelectable = true;
        view.drawsBackground = true

        return view as! Self;
    }
    
    /// 超链接处理
    /// - Parameter dic: [标题:网址]
    func hyperlink(dic: [String : String]) {
        let mattStr = NSAttributedString.hyperlink(dic: dic, text: self.string, font: self.font!)
        self.textStorage?.setAttributedString(mattStr)
        self.isEditable = false;
        self.isSelectable = true;
    }
    
}
